### ISSUE TITLE
NullReferenceException is thrown when using Fetch

### ISSUE THREAD

**Issue by Madajevas** [nhibernate/nhibernate-core#3288]
`NullReferenceException` in being thrown when trying to query with fetch and where condition on fetched data. This happens with v5.4.0+; v5.3.16 works without issues. Setup is a bit complex, hope that repro code bellow will explain it better.
```c#
using FluentNHibernate.Cfg.Db;
using FluentNHibernate.Cfg;
using NHibernate;
using NHibernate.Util;
using FluentNHibernate.Mapping;
using NHibernate.Linq;
using NUnit.Framework;

public class Tests
{
    private ISessionFactory sessionFactory;

    [SetUp]
    public void Setup()
    {
        var config = Fluently.Configure()
            .Database(MsSqlConfiguration.MsSql2008.ConnectionString("Server=localhost;Database=test;User Id=sa;Password=Test4000;"))
            .Mappings(m => m.FluentMappings.AddFromAssemblyOf<Tests>())
            .BuildConfiguration();

        sessionFactory = config.BuildSessionFactory();
    }

    [Test]
    public void Test1()
    {
        using var session = sessionFactory.OpenSession();

        var entities = session.Query<TopEntity>()
            .Fetch(e => e.MiddleEntity)
            .Where(e => e.MiddleEntity.BottomEntities.Any(e => e.OneMoreEntity != null))
            .ToList();
    }
}

class TopEntity
{
    public virtual int Id { get; set; }
    public virtual MiddleEntity MiddleEntity { get; set; }
}
class TopEntityMap : ClassMap<TopEntity>
{
    public TopEntityMap()
    {
        Id(e => e.Id);
        References(e => e.MiddleEntity);
    }
}

class MiddleEntity
{
    public virtual int Id { get; set; }
    public virtual ISet<BottomEntity> BottomEntities { get; set; }
}
class MiddleEntityMap : ClassMap<MiddleEntity>
{
    public MiddleEntityMap()
    {
        Id(e => e.Id);
        HasMany(e => e.BottomEntities)
            .Table("BottomEntity")
            .Cascade.All()
            .KeyColumn("MiddleEntity_id").Not.KeyNullable()
            .Component(e => {
                e.ParentReference(x => x.MiddleEntity);
                e.References(x => x.OneMoreEntity);
            });
    }
}

class BottomEntity
{
    public virtual MiddleEntity MiddleEntity { get; set; }
    public virtual OneMoreEntity OneMoreEntity { get; set; }

    public override bool Equals(object? obj)
    {
        return (obj as BottomEntity)?.MiddleEntity.Id == MiddleEntity.Id;
    }

    public override int GetHashCode()
    {
        return MiddleEntity.Id.GetHashCode();
    }
}
class BottomEntityMap : ClassMap<BottomEntity>
{
    public BottomEntityMap()
    {
        CompositeId()
            .KeyReference(x => x.MiddleEntity, "MiddleEntity_id")
            .KeyReference(x => x.OneMoreEntity, "OneMoreEntity_id");
    }
}

class OneMoreEntity
{
    public virtual int Id { get; set; }
}
class OneMoreEntityMap : ClassMap<OneMoreEntity>
{
    public OneMoreEntityMap()
    {
        Id(e => e.Id);
    }
}
```

Tested with the following database schema:
```sql
drop table if exists TopEntity;
create table TopEntity
(
	id int primary key,
	MiddleEntity_id int not null
);

drop table if exists MiddleEntity;
create table MiddleEntity
(
	id int primary key
);

drop table if exists BottomEntity;
create table BottomEntity
(
	MiddleEntity_id int not null,
	OneMoreEntity_id int not null
);

drop table if exists OneMoreEntity;
create table OneMoreEntity
(
	id int primary key
);
```

**Comment by bahusoid** [comment]
> This happens with v4.0+; v5.3.16 works without issues

v4 is no longer supported. Or did you mean 5.4? Does it fail with latest 5.4.2?

**Comment by bahusoid** [comment]
Something is not right with the mapping. Please adjust it so correct SQL is executed on not affected versions.

v5.3.16 generates invalid SQL:
```sql
SELECT topentity0_.Id AS id1_0_0_
	,middleenti2_.Id AS id1_2_1_
	,topentity0_.MiddleEntity_id AS middleentity2_0_0_
FROM "TopEntity" topentity0_
LEFT OUTER JOIN "MiddleEntity" middleenti1_ ON topentity0_.MiddleEntity_id = middleenti1_.Id
LEFT OUTER JOIN "MiddleEntity" middleenti2_ ON topentity0_.MiddleEntity_id = middleenti2_.Id
WHERE EXISTS (
		SELECT   -- WRONG SQL HERE.  EMPTY SELECT STATEMENT
		FROM BottomEntities bottomenti3_
		WHERE middleenti1_.Id = bottomenti3_.MiddleEntity_id
		)
```

**Comment by Madajevas** [comment]
It looks like I have removed too much code trying to make it bare minimum. Here is a more close resemblance to what we have and it works on v5.3.16 (and fails as before on v.5.4.0)

```csharp
using FluentNHibernate.Cfg.Db;
using FluentNHibernate.Cfg;
using NHibernate;
using NHibernate.Util;
using FluentNHibernate.Mapping;
using NHibernate.Linq;
using NUnit.Framework;

public class Tests
{
    private ISessionFactory sessionFactory;

    [SetUp]
    public void Setup()
    {
        var config = Fluently.Configure()
            .Database(MsSqlConfiguration.MsSql2008.ConnectionString("Server=localhost;Database=test;User Id=sa;Password=Test4000;"))
            .Mappings(m => m.FluentMappings.AddFromAssemblyOf<Tests>())
            .BuildConfiguration();

        sessionFactory = config.BuildSessionFactory();
    }

    [Test]
    public void Test1()
    {
        using var session = sessionFactory.OpenSession();

        var entities = session.Query<TopEntity>()
            .Fetch(e => e.MiddleEntity)
            .Where(e => e.MiddleEntity.BottomEntities.Any(e => e.OneMoreEntity != null))
            .ToList();
    }
}

class TopEntity
{
    public virtual int Id { get; set; }
    public virtual MiddleEntity MiddleEntity { get; set; }
}
class TopEntityMap : ClassMap<TopEntity>
{
    public TopEntityMap()
    {
        Id(e => e.Id);
        References(e => e.MiddleEntity);
    }
}

class MiddleEntity
{
    public virtual int Id { get; set; }
    public virtual ISet<BottomEntity> BottomEntities { get; set; }
}
class MiddleEntityMap : ClassMap<MiddleEntity>
{
    public MiddleEntityMap()
    {
        Id(e => e.Id);
        HasMany(e => e.BottomEntities)
            .Table("BottomEntity")
            .Cascade.All()
            .KeyColumn("MiddleEntity_id").Not.KeyNullable()
            .Component(e => {
                e.ParentReference(x => x.MiddleEntity);
                e.References(x => x.OneMoreEntity);
            });
    }
}

class BottomEntity
{
    public virtual MiddleEntity MiddleEntity { get; set; }
    public virtual OneMoreEntity OneMoreEntity { get; set; }

    public override bool Equals(object? obj)
    {
        return (obj as BottomEntity)?.MiddleEntity.Id == MiddleEntity.Id;
    }

    public override int GetHashCode()
    {
        return MiddleEntity.Id.GetHashCode();
    }
}
class BottomEntityMap : ClassMap<BottomEntity>
{
    public BottomEntityMap()
    {
        CompositeId()
            .KeyReference(x => x.MiddleEntity, "MiddleEntity_id")
            .KeyReference(x => x.OneMoreEntity, "OneMoreEntity_id");
    }
}

class OneMoreEntity
{
    public virtual int Id { get; set; }
}
class OneMoreEntityMap : ClassMap<OneMoreEntity>
{
    public OneMoreEntityMap()
    {
        Id(e => e.Id);
    }
}
```

Tested with the following database schema:
```sql
drop table if exists TopEntity;
create table TopEntity
(
	id int primary key,
	MiddleEntity_id int not null
);

drop table if exists MiddleEntity;
create table MiddleEntity
(
	id int primary key
);

drop table if exists BottomEntity;
create table BottomEntity
(
	MiddleEntity_id int not null,
	OneMoreEntity_id int not null
);

drop table if exists OneMoreEntity;
create table OneMoreEntity
(
	id int primary key
);
```

**Comment by fredericDelaporte** [comment]
It seems likely to me it has the same root cause than #3290. May you try disabling `detect_fetch_loops` and setting `max_fetch_depth` to an appropriate, large enough value for your case? See bef3bb7 `Configure` method.

**Comment by bahusoid** [comment]
> It seems likely to me it has the same root cause than https://github.com/nhibernate/nhibernate-core/issues/3290.

Nope. Another regression from #2496 revealing old hql bug

**Comment by bahusoid** [comment]
Fixed by #3314
